trin-core/src/utp/packets.rs

```diff
@@ -0,0 +1,322 @@
+        buf.extend_from_slice(&self.timestamp_difference.to_be_bytes());
+        buf.extend_from_slice(&self.wnd_size.to_be_bytes());
+        buf.extend_from_slice(&self.seq_nr.to_be_bytes());
+        buf.extend_from_slice(&self.ack_nr.to_be_bytes());
+        buf
+    }
+
+    pub fn decode(bytes: &[u8]) -> PacketHeader {
+        PacketHeader {
+            type_ver: bytes[0],
+            extension: bytes[1],
+            connection_id: u16::from_be_bytes([bytes[2], bytes[3]]),
+            timestamp: u32::from_be_bytes([bytes[4], bytes[5], bytes[6], bytes[7]]),
+            timestamp_difference: u32::from_be_bytes([bytes[8], bytes[9], bytes[10], bytes[11]]),
+            wnd_size: u32::from_be_bytes([bytes[12], bytes[13], bytes[14], bytes[15]]),
+            seq_nr: u16::from_be_bytes([bytes[16], bytes[17]]),
+            ack_nr: u16::from_be_bytes([bytes[18], bytes[19]]),
+        }
+    }
+
+    pub fn version(&self) -> u8 {
+        self.type_ver & 0xf
+    }
+    pub fn set_version(&mut self, int: u8) {
+        self.type_ver = (self.type_ver & 0xf0) | (int & 0xf)
+    }
+    pub fn type_(&self) -> PacketType {
+        PacketType::try_from(self.type_ver >> 4).unwrap()
+    }
+    pub fn set_type(&mut self, t: PacketType) {
+        self.type_ver = (self.type_ver & 0xf) | (u8::from(t) << 4)
+    }
+
+    pub fn set_connection_id(&mut self, connection_id: u16) {
+        self.connection_id = connection_id;
+    }
+
+    pub fn set_timestamp(&mut self, timestamp: u32) {
+        self.timestamp = timestamp;
+    }
+
+    pub fn set_timestamp_difference(&mut self, timestamp_difference: u32) {
+        self.timestamp_difference = timestamp_difference;
+    }
+
+    pub fn set_wnd_size(&mut self, wnd_size: u32) {
+        self.wnd_size = wnd_size;
+    }
+
+    pub fn set_seq_nr(&mut self, seq_nr: u16) {
+        self.seq_nr = seq_nr;
+    }
+
+    pub fn set_ack_nr(&mut self, ack_nr: u16) {
+        self.ack_nr = ack_nr;
+    }
+}
+
+#[derive(Debug)]
+pub struct Extension {
+    pub extension_type: u8,
+    pub len: u8,
+    pub bitmask: Vec<u8>,
+}
+
+#[derive(Clone, Debug)]
+pub struct Packet(pub Vec<u8>);
+
+impl Packet {
+    pub fn new(mut packet_header: PacketHeader) -> Packet {
+        let mut vec = Vec::with_capacity(HEADER_SIZE);
+        vec.append(&mut packet_header.encode());
+        Packet(vec)
+    }
+
+    pub fn with_payload(mut packet_header: PacketHeader, payload: &[u8]) -> Packet {
+        let mut vec = Vec::with_capacity(HEADER_SIZE + payload.len());
+        vec.append(&mut packet_header.encode());
+        vec.extend_from_slice(payload);
+        Packet(vec)
+    }
+
+    pub fn from_slice(bytes: &[u8]) -> Self {
+        Self(bytes.to_vec())
+    }
+
+    pub fn get_header(&self) -> PacketHeader {
+        PacketHeader::decode(&self.0[0..20])
+    }
+
+    pub fn type_(&self) -> PacketType {
+        self.get_header().type_()
+    }
+
+    pub fn version(&self) -> u8 {
```

Comment:
What is the reason to have all of those helper methods in the `Packet`? If `Packet` is composed of `PacketHeader` and `PacketHeader` is composed of `type`, `ver`, `extension`, etc, then we can just call `packet.header.version` etc..

Reply:
It is cleaner to look at in code if we have to call packet.header.version 10 times. packet.version() is a lot nicer to read.

Reply:
> It is cleaner to look at in code if we have to call packet.header.version 10 times. packet.version() is a lot nicer to read.

If this is the case, I think we can implement the [Deref](https://doc.rust-lang.org/std/ops/trait.Deref.html) trait and call `packet.version()`.